src/Metalama.Patterns.NotifyPropertyChanged/Implementation/Natural/BuildAspectContext.cs

```diff
@@ -0,0 +1,451 @@
+﻿// Copyright (c) SharpCrafters s.r.o. See the LICENSE.md file in the root directory of this repository root for details.
+
+using Metalama.Framework.Aspects;
+using Metalama.Framework.Code;
+using Metalama.Framework.Code.Collections;
+using Metalama.Framework.Diagnostics;
+using Metalama.Framework.Engine.Diagnostics;
+using Metalama.Patterns.NotifyPropertyChanged.Metadata;
+using System.ComponentModel;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Metalama.Patterns.NotifyPropertyChanged.Implementation.Natural;
+
+[CompileTime]
+internal sealed partial class BuildAspectContext
+{
+    private static readonly string[] _onPropertyChangedMethodNames = { "OnPropertyChanged", "NotifyOfPropertyChange", "RaisePropertyChanged" };
+
+    private readonly Dictionary<IType, InpcInstrumentationKind> _inpcInstrumentationKindLookup = new();
+    private readonly Dictionary<IFieldOrProperty, bool> _validateFieldOrPropertyLookup = new();
+    private readonly Lazy<IMethod?> _baseOnPropertyChangedMethod;
+    private readonly Lazy<IMethod?> _baseOnChildPropertyChangedMethod;
+    private readonly Lazy<IMethod?> _baseOnUnmonitoredObservablePropertyChangedMethod;
+
+    private HashSet<string>? _inheritedOnChildPropertyChangedPropertyPaths;
+    private HashSet<string>? _inheritedOnUnmonitoredObservablePropertyChangedPropertyNames;
+
+    /// <summary>
+    /// Gets frequently used compilation elements resolved for the current compilation.
+    /// </summary>
+    public ElementsRecord Elements { get; }
+
+    public BuildAspectContext( IAspectBuilder<INamedType> builder )
+    {
+        this.Elements = new ElementsRecord();
+        this.Builder = builder;
+
+        var target = builder.Target;
+
+        // TODO: Consider using BaseType.Definition where possible for better performance.
+
+        this.BaseImplementsInpc =
+            target.BaseType != null && (
+                target.BaseType.Is( this.Elements.INotifyPropertyChanged )
+                || (target.BaseType is { BelongsToCurrentProject: true }
```

Comment:
For posterity: I believe that the condition `&& (!this.Target.Compilation.IsPartial || this.Target.Compilation.Types.Contains( target.BaseType 
))` was added to deal with aspect tests where the base class is not in the same file as the test class - in which case the compilation is in effect partial although not `IsPartial`. Those tests ended up being refactored to have the base class in the same file as the test class, thus avoiding the problem. So that condition can now be removed.